planning_scene/src/planning_scene.cpp

```diff
@@ -1245,60 +1309,101 @@ bool planning_scene::PlanningScene::usePlanningSceneMsg(const moveit_msgs::Plann
-  if (!map.header.frame_id.empty())
+void planning_scene::PlanningScene::processOctomapMsg(const octomap_msgs::Octomap &msg, const Eigen::Affine3d &t)
+{
+  if (msg.id.empty())
   {
-    const Eigen::Affine3d &t = getTransforms().getTransform(map.header.frame_id);
+    world_->removeObject(OCTOMAP_NS);
+  }
+  else if (msg.id == OCTOMAP_DIFF_MSG_TYPE)
+  {
+    collision_detection::CollisionWorld::ObjectConstPtr map = world_->getObject(OCTOMAP_NS);
+    if (map && map->shapes_.size() == 1)
+    {
+      if (msg.data.empty())
+        return;
+      const shapes::OcTree *o = static_cast<const shapes::OcTree*>(map->shapes_[0].get());
+      boost::shared_ptr<const octomap::OcTree> octree = o->octree;
+      boost::shared_ptr<octomap::OcTree> nc_octree = boost::const_pointer_cast<octomap::OcTree>(octree);
+      processOctomapMsgDiff(msg, nc_octree);
+      processOctomapPtr(octree, t);
+    }
+  }
+  else if (msg.id == OCTOMAP_MSG_TYPE)
+  {
+    world_->removeObject(OCTOMAP_NS); // Octomap replaces any previous one
+    if (msg.data.empty())
+      return;
+    boost::shared_ptr<octomap::OcTree> om(static_cast<octomap::OcTree*>(octomap_msgs::msgToMap(msg)));
     world_->addToObject(OCTOMAP_NS, shapes::ShapeConstPtr(new shapes::OcTree(om)), t);
   }
   else
   {
-    world_->addToObject(OCTOMAP_NS, shapes::ShapeConstPtr(new shapes::OcTree(om)), Eigen::Affine3d::Identity());
+    logError("Received Octomap is of unknown type '%s'", msg.id.c_str());
   }
 }
 
-void planning_scene::PlanningScene::removeAllCollisionObjects()
+void planning_scene::PlanningScene::processOctomapMsgDiff(const octomap_msgs::Octomap &msg, boost::shared_ptr<octomap::OcTree> octree)
 {
-  const std::vector<std::string> &object_ids = world_->getObjectIds();
-  for (std::size_t i = 0; i < object_ids.size(); ++i)
-    if (object_ids[i] != OCTOMAP_NS)
-      world_->removeObject(object_ids[i]);
-}
-
-void planning_scene::PlanningScene::processOctomapMsg(const octomap_msgs::OctomapWithPose &map)
-{
-  // each octomap replaces any previous one
-  world_->removeObject(OCTOMAP_NS);
+  std::stringstream datastream;
+  datastream.write((const char*) &msg.data[0], msg.data.size());
 
-  if (map.octomap.data.empty())
-    return;
-
-  if (map.octomap.id != "OcTree")
+  int num_changes;
+  datastream.read((char*) &num_changes, sizeof(int));
+  //logInform("Octomap diff has %i changes", num_changes);
+  int expected_size = sizeof(int)+num_changes*((3*sizeof(unsigned short int))+sizeof(float));
+  if (expected_size > msg.data.size())
   {
-    logError("Received octomap is of type '%s' but type 'OcTree' is expected.", map.octomap.id.c_str());
+    logError("Did not receive enough data for specified diff size: %i bytes expected, %i received", expected_size, msg.data.size());
     return;
   }
+  if(expected_size < msg.data.size()) {
+      logWarn("Got more data than expected (%zu > %d)", msg.data.size(), expected_size);
+  }
```

Comment:
Actually that seems reasonable. If the data size is not what is expected (either less or more) something is probably wrong.

I would not go for anything that aborts without the option to only enable this globally. This could destroy a working system because of "something".